docs/etcd.md

```diff
@@ -0,0 +1,405 @@
+- Combination of basic operations and transactions into an atomic operation with multiple read phases, see [AtomicOp](#atomicop).
+- Connection of a processor/callback to each operation and its parts.
+- [Iterator](#iterator) loads all pages from the same revision snapshot.
+- High-level Watch API streams - with manual restarts and automatic restarts on network failure, see [Watch Streams](#watch-streams).
+- High-level sessions/leases - with retries on network failure, see [Session](#session).
+
+### Packages
+
+The `etcdop` package consists of the following sub-packages:
+- [etcdop](../internal/pkg/service/common/etcdop) - prefix, key, watch stream.
+- [etcdop/serde](../internal/pkg/service/common/etcdop/serde), see [Serde](#serde).
+- [etcdop/op](../internal/pkg/service/common/etcdop/op) - basic operations, [TxnOp](#txnop), [AtomicOp](#atomicop).
+- [etcdop/iterator](../internal/pkg/service/common/etcdop/iterator), see [Iterator](#iterator).
+
+**In the future, it would be beneficial to combine these sub-packages into a single `etcdop` package, possibly excluding `serde`.**
+
+### Serde
+
+The `serde` package provides encoding, decoding, and validation operations for any value stored in etcd.
+
+Entities are validated upon save, but also upon load.
+
+The `serde.Serde` is an argument of the `etcdop.NewTypedKey` and `etcdop.NewTypedPrefix` constructors.
+
+Supported encodings:
+- Currently, only the `JSON` encoding is implemented.
+- If necessary in the future, it is possible to implement additional encodings.
+  - For example, for performance or database size optimization.
+- Possible extensions could include `gzip` compression or binary [Protocol Buffers](https://protobuf.dev) encoding.
+- It is possible to work with several encodings simultaneously:
+  - This allows for maintaining backward compatibility.
+  - Or implementing gradual migration.
+  - For example, a header/prefix can be used to determine encoding:
+    - `{` or `[` for JSON
+    - `PB:` for Protocol Buffers
+    - `GZIP:` for compression.
+  - The disadvantage is more complicated debugging, as values will not be directly readable.
+
+### Schema
+
+When using the `etcdop` framework, we must start with a schema.
+
+The schema defines:
+- Prefixes:
+  - `etcdop.Prefix` type is for raw values, represented by bytes.
+  - `etcdop.PrefixT[T]` type is for values encoded/decoded by the [serde](#serde).
+- Keys:
+  - `etcdop.Key` type is for raw values, represented by bytes.
+  - `etcdop.KeyT[T]` type is for values encoded/decoded by the [serde](#serde).
+
+Example of defining a typed prefix and key:
+```go
+s := serde.NewJSON(validator.Validate)
+pfx1 := etcdop.NewTypedPrefix[model.Slice]("storage/slice", s)
+pfx2 := pfx1.Add("level").Add("local")
+key1 := pfx2.Key(slice1.Key())
+key2 := pfx2.Key(slice2.Key())
+require.NoError(t, key1.Put(slice1).Do(ctx).Err())
+require.NoError(t, key2.Put(slice2).Do(ctx).Err())
```

Comment:
I miss usage of  `GetAll` or any `get` or `listing` operation. Could you provide it somehow? :slightly_smiling_face: . When I tried to implement the storage cleaner instead of using the repository I used schema but it was unconvintient.
The documenation does not say so but we prefer to go through `repository` when accessing/manipulating with models/schemas.

Reply:
Ok, I will add `GetAll`.